### DISCUSSION TITLE
Periodically timed functions and waypoint arrival

### DISCUSSION THREAD

**Discussion by OrxataEnFartons** [TUDelft-CNS-ATM/bluesky#411] [Q&A] [answered]
Hi, 
I created in my plugin a timed function to forward the position of the drone to a server. It works, but I would prefer that timed function to start only when the aircraft starts flying (not just after the scenario is loaded), and stop it once the aircraft reaches its destination. Is there a way to stop calling a specific timed function?

Now, when an aircraft reaches the last point of the scenario it just follows with the current heading. Is there built any way to identify if an aircraft reached its destination, or to know the past/next point in the trajectory? I see nothing like that in the Traffic object. 
Is there maybe some option to quit an scenario when the aircraft reaches its destination?

Regards.

**Comment by OrxataEnFartons** [comment]
Well, somehow I managed to detect when it arrives to the end. Using the traf.actwp.swlastwp[idx] that becomes true when reaches to the last leg and when that happens, checking the distance to destination with each new position. When is less than a certain margin , I delete the aircraft with a DEL acid.

**Comment by jooste** [comment, marked as answer]
Hi @OrxataEnFartons,

You could also do this with stack commands. For instance, the following:
```
    CRE KL001 B738 52 4 0 FL250 250
    KL001 ADDWPT 53 4
    KL001 AT KL001001 DO DEL KL001
```
would call the stack command `DEL KL001` once the waypoint is reached, and thus delete the aircraft. You can let it call any stack command this way, including the ones you create yourself.

Timed functions run when the simulation is running, but you can for instance add a condition check in your function to skip the remainder of the function when there is no traffic:
```python
import bluesky as bs

@timed_function
def update():
     if bs.traf.ntraf == 0:
        return
```

**Reply by OrxataEnFartons** [reply]
Thanks. I tried that. I download the scn from a server, with lat lon points, so to pass the DEST to AT  (since this does not accept lat lon) I define its name first. 
But it doesn't work, the aircraft continues after reaching the last point. Not sure if AT has problems with newly defined wp names.
But well, the mentioned workaround does the work by now. 

00:00:00.00>DEFWPT M6001_DEST,52.306783,10.523382,FIX
00:00:00.00>DEFWPT M6001_WP0,52.231887,10.528944,FIX
00:00:00.00>DEFWPT M6001_WP1,52.237018,10.586552,FIX
00:00:00.00>DEFWPT M6001_WP2,52.245808,10.467696,FIX
00:00:00.00>DEFWPT M6001_WP3,52.257605,10.567834,FIX
00:00:00.00>DEFWPT M6001_WP4,52.26998,10.498063,FIX
00:00:00.00>DEFWPT M6001_WP5,52.278612,10.550374,FIX
00:00:00.00>DEFWPT M6001_WP6,52.289141,10.500978,FIX
00:01:33.00>CRE M6001,M600,52.22023,10.533871,0,0.0,20
00:01:33.00>ORIG,M6001,52.22023,10.533871
00:01:33.00>DEST,M6001,M6001_DEST
00:01:33.00>ADDWPT,M6001,M6001_WP0,98.4,38.88
00:01:33.00>ADDWPT,M6001,M6001_WP1,98.4,38.88
00:01:33.00>ADDWPT,M6001,M6001_WP2,98.4,38.88
00:01:33.00>ADDWPT,M6001,M6001_WP3,98.4,38.88
00:01:33.00>ADDWPT,M6001,M6001_WP4,98.4,38.88
00:01:33.00>ADDWPT,M6001,M6001_WP5,98.4,38.88
00:01:33.00>ADDWPT,M6001,M6001_WP6,98.4,38.88
00:01:33.00>LNAV,M6001,ON
00:01:33.00>VNAV,M6001,ON
00:01:33.00>M6001 AT M6001_DEST DO DEL M6001

**Reply by jooste** [reply]
Just to come back to this: When I run the example scenario you give here, the aircraft is correctly deleted when reaching `M6001_DEST`. I've also tested different examples, with `DEFWPT` and lat/lon waypoints, they all work with the `AT` command. If your version of BlueSky is not up to date with master you could try if updating from master solves the issues you mention.

**Reply by OrxataEnFartons** [reply]
At the moment I solved this by checking traf.actwp.swlastwp[idx] to see if it is already in the last leg. Then when I detect that I passed the destination point, I DEL the aircraft.

**Comment by ProfHoekstra** [comment]
Dear Miquel,

Ok, good to hear. Normally this should indeed be covered by the AT command,. Please check the possibilities of ther AT command, using the examples. You can send any command when the waypoint reaches the waypoint. There also other variants of the AT-command, called ATDIST, ATALT ATSPD etc, these could all help as well. We use AT and these commands regularly to program  specific logic or other conditional functions in the scenario. Why doesn’t it do what you wanted? What did you try to use it for exactly?

Best regards,
    Jacco Hoekstra

**Reply by OrxataEnFartons** [reply]
Hi Jacco, 
In my plugin I receive flight plans that have been designed clicking on a map. So the first thing I do when creating the scn is naming these waypoints:
00:00:00.00>DEFWPT M6001_WP3,52.257605,10.567834,FIX
Having these waypoints defined like this in the scn, and the scenario running, if the plugin executes a: 
AT M6001,M6001_WP3 ALT 32.808399, 
I didn't get an error but the altitude did not change when arriving that waypoint.
Same if I introduce  that command in the console while executing the scn. Maybe some problem recognizing the named waypoint? That is why I tried editing and reloading the whole scn.

**Comment by ProfHoekstra** [comment]
That is indeed strange, like I said, I will have a look  at this. Might indeed be that some exception is triggered due to the new SWTOD/SWTOC logic that has been added. Will fix it this week.
